### PR TITLE
Implement rules config page and add userdata webserver field

### PR DESCRIPTION
**Pull request by CurlyMoo**
In preperation of the rules functionality

### CODE REVIEW

**Comment by CurlyMoo** [comment]
@TD-er I moved the logic to a separate PR as i should've initially :smile:

File: HeishaMon/HeishaMon.ino [hunk 1]
```diff
@@ -652,6 +687,21 @@ int8_t webserver_cb(struct webserver_t *client, void *dat) {
                 return showFirmwareFail(client);
               }
             } break;
+          case 160: {
+              return showRules(client);
+            } break;
```
Comment (TD-er) [inline HeishaMon/HeishaMon.ino:692]: No need for `break;` after a return. Is unreachable code :)

Comment (CurlyMoo) [reply HeishaMon/HeishaMon.ino:692]: True, just a matter of automatism.

File: HeishaMon/src/common/webserver.h [hunk 2]
```diff
@@ -112,6 +112,7 @@ typedef struct webserver_t {
   webserver_cb_t *callback;
   unsigned char buffer[WEBSERVER_BUFFER_SIZE];
   char *boundary;
+  void *userdata;
```
Comment (TD-er) [inline HeishaMon/src/common/webserver.h:115]: Shouldn't these pointers be initialized to a `null_ptr` ?
```c++
  webserver_cb_t *callback = null_ptr;
  unsigned char buffer[WEBSERVER_BUFFER_SIZE] = {0};
  char *boundary = null_ptr;
  void *userdata = null_ptr;
```
(just picking what is shown in the diff, maybe other members too)

Comment (CurlyMoo) [reply HeishaMon/src/common/webserver.h:115]: Good point. They are in the other commit.

File: HeishaMon/webfunctions.cpp [hunk 3]
```diff
@@ -1034,12 +1026,87 @@ int handleJsonOutput(struct webserver_t *client, String actData[]) {
...
+    webserver_send_content_P(client, showRulesPage1, strlen_P(showRulesPage1));
+    if (LittleFS.begin()) {
+      client->userdata = new fs::File(LittleFS.open("/rules.txt", "r"));
+    }
+  } else if (client->userdata != NULL) {
+    #define BUFFER_SIZE 128
+    File *f = (File *)client->userdata;
+    char content[BUFFER_SIZE];
+    memset(content, 0, BUFFER_SIZE);
+    if (f) {
```
Comment (TD-er) [inline HeishaMon/webfunctions.cpp:1046]: Maybe also check if it is a valid file handle.
`FS::File` implements the bool operator for this.

`if (f && *f) {`

File: HeishaMon/webfunctions.h [hunk 4]
```diff
@@ -46,6 +46,12 @@ struct settingsStruct {
   gpioSettingsStruct gpioSettings;
 };
 
+struct websettings_t {
+  String name;
+  String value;
+  struct websettings_t *next;
```
Comment (TD-er) [inline HeishaMon/webfunctions.h:52]:
```c++
struct websettings_t *next = nullptr;
```

Comment (CurlyMoo) [reply HeishaMon/webfunctions.h:52]: Also done at initializiation.

**Comment by IgorYbema** [comment]
In v3.0 release

**Pull request closed without merging**